sql/core/src/main/scala/org/apache/spark/sql/execution/script/ScriptTransformationExec.scala

```diff
@@ -0,0 +1,239 @@
+        try {
+          if (curLine == null) {
+            curLine = reader.readLine()
+            if (curLine == null) {
+              checkFailureAndPropagate(writerThread, null, proc, stderrBuffer)
+              return false
+            }
+          }
+          true
+        } catch {
+          case NonFatal(e) =>
+            // If this exception is due to abrupt / unclean termination of `proc`,
+            // then detect it and propagate a better exception message for end users
+            checkFailureAndPropagate(writerThread, e, proc, stderrBuffer)
+
+            throw e
+        }
+      }
+
+      override def next(): InternalRow = {
+        if (!hasNext) {
+          throw new NoSuchElementException
+        }
+        val prevLine = curLine
+        curLine = reader.readLine()
+        if (!ioschema.isSchemaLess) {
+          new GenericInternalRow(
+            prevLine.split(ioschema.outputRowFormatMap("TOK_TABLEROWFORMATFIELD"))
+              .map(CatalystTypeConverters.convertToCatalyst))
+        } else {
+          new GenericInternalRow(
+            prevLine.split(ioschema.outputRowFormatMap("TOK_TABLEROWFORMATFIELD"), 2)
+              .map(CatalystTypeConverters.convertToCatalyst))
+        }
+      }
+    }
+
+    writerThread.start()
+
+    outputIterator
+  }
+}
+
+private class ScriptTransformationWriterThread(
+    iter: Iterator[InternalRow],
+    inputSchema: Seq[DataType],
+    ioschema: ScriptTransformIOSchema,
+    outputStream: OutputStream,
+    proc: Process,
+    stderrBuffer: CircularBuffer,
+    taskContext: TaskContext,
+    conf: Configuration)
+  extends ScriptTransformationWriterThreadBase(
+      iter,
+      inputSchema,
+      outputStream,
+      proc,
+      stderrBuffer,
+      taskContext,
+      conf) {
+
+  setDaemon(true)
+
+  protected val lineDelimiter = ioschema.inputRowFormatMap("TOK_TABLEROWFORMATLINES")
+  protected val fieldDelimiter = ioschema.inputRowFormatMap("TOK_TABLEROWFORMATFIELD")
+
+  override def processRows(): Unit = {
+    val len = inputSchema.length
+    iter.foreach { row =>
+      val data = if (len == 0) {
+        ioschema.inputRowFormatMap("TOK_TABLEROWFORMATLINES")
+      } else {
+        val sb = new StringBuilder
+        sb.append(row.get(0, inputSchema(0)))
+        var i = 1
+        while (i < len) {
+          sb.append(ioschema.inputRowFormatMap("TOK_TABLEROWFORMATFIELD"))
+          val columnType = inputSchema(i)
+          val fieldValue = row.get(i, columnType)
+          val fieldStringValue = columnType match {
+            case _: DateType =>
+              val dateFormatter = DateFormatter(DateTimeUtils.defaultTimeZone.toZoneId)
+              dateFormatter.format(fieldValue.asInstanceOf[Int])
+            case _: TimestampType =>
+              DateTimeUtils.timestampToString(
+                TimestampFormatter.getFractionFormatter(DateTimeUtils.defaultTimeZone.toZoneId),
+                fieldValue.asInstanceOf[SQLTimestamp])
```

Comment:
Two questions I have here;
 - Why do we need the special handling for date/timestamp here? It seems the original one (w/ the null serde case) simply output as strings: https://github.com/apache/spark/blob/6befb2d8bdc5743d0333f4839cf301af165582ce/sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformationExec.scala#L289-L306
 - Do you have any plan to support a custom serde in the Spark-native implementation?

Reply:
> * Why do we need the special handling for date/timestamp here? It seems the original one (w/ the null serde case) simply outputs rows as strings though:

For this question, if we remove code about handle DataTyle/TimestampType, for test case `SPARK-25990: TRANSFORM should handle different data types correctly` will get below error, since  https://github.com/apache/spark/pull/14279
```
[info] - SPARK-25990: TRANSFORM should handle different data types correctly *** FAILED *** (4 seconds, 997 milliseconds)
[info]   Results do not match for Spark plan:
[info]    ScriptTransformation [a#19, b#20, c#21, d#22, e#23, f#24], python /Users/angerszhu/Documents/project/AngersZhu/spark/sql/core/target/scala-2.12/test-classes/test_script.py, [a#31, b#32, c#33, d#34, e#35, f#36], org.apache.spark.sql.execution.script.ScriptTransformIOSchema@1ad5a29c
[info]   +- Project [_1#6 AS a#19, _2#7 AS b#20, _3#8 AS c#21, _4#9 AS d#22, _5#10 AS e#23, _6#11 AS f#24]
[info]      +- LocalTableScan [_1#6, _2#7, _3#8, _4#9, _5#10, _6#11]
[info]
[info]
[info]    == Results ==
[info]    !== Expected Answer - 3 ==                                           == Actual Answer - 3 ==
[info]   ![1,1,1.0,1.000000000000000000,1970-01-01 08:00:00.001,2015-05-21]   [1,1,1.0,1.000000000000000000,1000,16576]
[info]   ![2,2,2.0,2.000000000000000000,1970-01-01 08:00:00.002,2015-05-22]   [2,2,2.0,2.000000000000000000,2000,16577]
[info]   ![3,3,3.0,3.000000000000000000,1970-01-01 08:00:00.003,2015-05-23]   [3,3,3.0,3.000000000000000000,3000,16578] (SparkPlanTest.scala:95)
[
```

Reply:
> * Why do we need the special handling for date/timestamp here? It seems the original one (w/ the null serde case) simply outputs rows as strings though:

For this part, maybe I should make a new jira as a sub-task of this?

Reply:
> * Do you have any plan to support a custom serde in the Spark-native implementation?

We are discusses about this, if we need to still keep serde, if need, I will   do this work.

Reply:
@maropu 
Remove logical about handle DataType/TimestampType and add a new sub-task, fix this after this PR